### ISSUE TITLE
Roadmap for the Laravel - v4

### ISSUE THREAD

**Issue by sylvainar** [ampache/ampache#1683]
Hi,

Is it possible to have more transparency about what are your progresses in the v4? Do you have a roadmap of what is already done and what's left?

Thanks !

**Comment by wagnered** [comment]
@sylvainar 
Thanks for your interest.

Being the only active developer on the project makes it difficult  developing the Ampache/Laravel version and solving user issues at the same time.  V4 is a considerable learning process being new to Laravel, w3-CSS, jQuery, etc. so a "roadmap" would be difficult to lay out.  Also, the fact that 3.8.* is so very unorganized it is rather difficult "untangling" the code and adapting it to a framework. It's far from dropping in the script and expecting it to be without errors.

A user can register and log in. I am adding roles for more versatility. I started working on the catalog/module creation procedures. This has to be done before completing other aspects of the design.  Also, there are  all the users enhancement requests  to be considered and configured in if possible.

**Comment by a7medo778** [comment]
To be honest, i think a complete from scratch approach is a bit too harsh, we can do with a facelift of the UI and making it more mobile friendly, and frankly for 99% of the users that would be it.

Now if we to be performance centric, the api needs rephrasing, i been running madsonic vs ampache for a while now, and the response time / load endurance is so vast between the two

If the api is fast i will take the project my self of creating an angular front end that non admin users can use

**Comment by wagnered** [comment]
@a7medo778 

> we can do with a facelift of the UI and making it more mobile friendly,

 And that would take a major rework : newer css bootstrap, jquery, etc.

> and frankly for 99% of the users that would be it.

I don't think so. :smile: 

As I understand, Madsonic, like Airsonic, is a "pre-proprietory" fork of Subsonic.

I'm sure java is faster than php, but not as dynamic. Many Ampache aficionados, like Ampache as a front end for the Subsonic type server, because of the access to many enhancements  that the Subsonic type server doesn't provide.

Many of the requested enhancements would require a major rework of the existing Ampache logic. Why not rebuild it on a framework that would allow more rapid addition of enhancements?

Laravel provides both Vue.js and React.js for the front-end scaffolding.  Either one would be adequate.

**Comment by a7medo778** [comment]
Thanks for the reply ernest, i guess laravel it is then

**Comment by a7medo778** [comment]
hey @wagnered 
i am building my next app based on ampache (ditching madsonic for opensource), will current versions be upgradable to v4 ? 

and in a related question, does php7 improves the current version performance ? or it doesnt make a diff since it was built for php5

**Comment by wagnered** [comment]
@a7medo778 
>will current versions be upgradable to v4 ?

The Ampache subsonic api section is "transparent" to any Subsonic client. If your app can issue all Subsonic commands, it should be ready for any Ampache Subsonic API upgrade. For example, Ultrasonic android client recently modified their app to require json responses instead of XML responses. I modified the Ampache Subsonic API to meet the Ultrasonic requirements and continue to be backward compatible with XML.

Ampache Subsonic API will continue to include more Subsonic commands with the API: The challenge to you will be to be to make sure your app can issue all Subsonic api commands.

>does php7 improves the current version performance ?

I did a search of 5.6 vs 7 and it seems that with few users, they are about the same,  with 200 + users, 7.0 seems faster.  I haven't done any performance tests myself.

**Comment by a7medo778** [comment]
yeah thats for sure, but i guess for now i just wanted to make sure that
its actually working
i hope the client i build get some attention into ampache

**Comment by a7medo778** [comment]
@wagnered hey wagnered, sorry for asking again, but would the DB schema change ? i need to pull some stuff directly from the db, should i expect a change in the laravel version ? (and how is the progress there?)

**Comment by wagnered** [comment]
@a7medo778 

Yes, there will be additional columns in some tables and also new tables.  If you intend to bypass  the Ampache functionality and access only the metadata directly, there doesn't seem to be any changes **(yet)**.  The new tables and table columns are for catalog management.  I haven't started on creating the catalogs at this point in time so the data captured for creating catalogs might evolve..

How is the progress?  Slow.  Remember, The "classic" Ampache has evolved over many years  A lot of little details to unravel and "translate" into the Laravel framework.  Also, it takes time to determine if requested enhancements can be incorporated into the new design.

But I am making progress and am enjoying the tasks.

**Comment by a7medo778** [comment]
@wagnered much appreciated, i guess i can always update the version when the laravel version is out

catalogs scanning specially requires ALOT of work since the current performance is miserable. keep up the awesome work

**Comment by a7medo778** [comment]
ey @wagnered , how is the laravel version going :P

**Comment by wagnered** [comment]
I've got about 95% of user management completed including login and registration. It also includes captcha and email verification.  My understanding of Laravel is accelerating: Many concepts in the documentation have finally become clear to me, as I finally have a need to  actually incorporate the concepts, but there are many more to go.

I've spent two or more weeks trying to use JQuery UI for many of the dialogs without much success, so I reverted back to responding with html pages. 

Much time is utilized attempting to find the best way to adapt a feature from the old version to the Laravel Framework.  I've worked in engineering environments and some projects might take a year or more to release alpha versions.  That's with a full team.  It will probably take me at least a year working by myself, to release an alpha version of Ampache/Laravel.

**Comment by a7medo778** [comment]
Wow that sounds awesome 😆
I think once the initial beta is out maybe more contributors will be attracted to pull some fixes.
Anyways, great job, looking forward for some performance godness

**Comment by a7medo778** [comment]
Hey @wagnered
Any updates 😅, we need something to build on top

**Comment by wagnered** [comment]
> I think once the initial beta is out maybe more contributors will be attracted to pull some fixes.

They would have to have a solid understanding of Laravel and the MVC architecture. MVC is a software architecture pattern and it stands for Model View Controller.  And I am finally getting (almost) adequately proficient.  One of the major problems I have is that all my learning is done by continual web searches as the only other developer who might be familiar with Laravel is pre-occupied by work of all things. :frowning_face: 

I've spent a lot of time trying to create an Ampache initialization similar to the current version, without success, so I opted for using Artisan, a command line feature to query the user and install the database, user, etc. 

It would be nice if someone knew the current version well enough and has time to volunteer fixes.  This would leave me more time to work on V4. :smile: 

I had a setback: My hard drive died and I had to replace it and rebuild the system.  I finally got the last of the bugs out of the new system build and what was carried over from the old system.  For example I had to downgrade Firefox to version 59 because Eclipse, my design IDE, couldn't display a web page when starting a debug session with the latest version of Firefox.

Hopefully, in the (relatively) near future, I will be at a point that the branch can be forked and demo the init,  register, user add and maybe a few more functions.

**Comment by a7medo778** [comment]
> > I think once the initial beta is out maybe more contributors will be attracted to pull some fixes.
> 
> They would have to have a solid understanding of Laravel and the MVC architecture. MVC is a software architecture pattern and it stands for Model View Controller. And I am finally getting (almost) adequately proficient. One of the major problems I have is that all my learning is done by continual web searches as the only other developer who might be familiar with Laravel is pre-occupied by work of all things. 
> 
> I've spent a lot of time trying to create an Ampache initialization similar to the current version, without success, so I opted for using Artisan, a command line feature to query the user and install the database, user, etc.
> 
> It would be nice if someone knew the current version well enough and has time to volunteer fixes. This would leave me more time to work on V4. 
> 
> I had a setback: My hard drive died and I had to replace it and rebuild the system. I finally got the last of the bugs out of the new system build and what was carried over from the old system. For example I had to downgrade Firefox to version 59 because Eclipse, my design IDE, couldn't display a web page when starting a debug session with the latest version of Firefox.
> 
> Hopefully, in the (relatively) near future, I will be at a point that the branch can be forked and demo the init, register, user add and maybe a few more functions.

Ok i was hoping that someone will jump in but i guess not 😂
Anyways, so i have a couple of php devs that are helping me add the youtube channel thingy
https://github.com/ampache/ampache/issues/1796
Should we work on this codebase, or keep waiting for the laravel?

**Comment by wagnered** [comment]
@a7medo778 

I'll say it this way:  If  you want to keep working on it, I'll mark the PR as an enhancement to be incorporated into Ampache/Laravel.

**Comment by lachlan-00** [comment]
closing as laravel is not being worked on now.
